src/openzeppelin/token/erc1155/library.cairo

```diff
@@ -0,0 +1,601 @@
+//
+
+@storage_var
+func ERC1155_balances(id: Uint256, account: felt) -> (balance: Uint256) {
+}
+
+@storage_var
+func ERC1155_operator_approvals(account: felt, operator: felt) -> (approved: felt) {
+}
+
+@storage_var
+func ERC1155_uri() -> (uri: felt) {
+}
+
+namespace ERC1155 {
+    //
+    // Initializer
+    //
+
+    func initializer{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(uri: felt) {
+        _set_uri(uri);
+        ERC165.register_interface(IERC1155_ID);
+        ERC165.register_interface(IERC1155_METADATA_ID);
+        return ();
+    }
+
+    //
+    // Modifiers
+    //
+
+    func assert_owner_or_approved{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        owner
+    ) {
+        let (caller) = get_caller_address();
+        if (caller == owner) {
+            return ();
+        }
+        let (approved) = ERC1155.is_approved_for_all(owner, caller);
+        with_attr error_message("ERC1155: caller is not owner nor approved") {
+            assert approved = TRUE;
+        }
+        return ();
+    }
+
+    //
+    // Getters
+    //
+
+    func uri{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(id: Uint256) -> (
+        uri: felt
+    ) {
+        let (uri) = ERC1155_uri.read();
+        return (uri,);
+    }
+
+    func balance_of{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        account: felt, id: Uint256
+    ) -> (balance: Uint256) {
+        with_attr error_message("ERC1155: address zero is not a valid owner") {
+            assert_not_zero(account);
+        }
+        _check_id(id);
+        let (balance) = ERC1155_balances.read(id, account);
+        return (balance,);
+    }
+
+    func balance_of_batch{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        accounts_len: felt, accounts: felt*, ids_len: felt, ids: Uint256*
+    ) -> (batch_balances_len: felt, batch_balances: Uint256*) {
+        alloc_locals;
+        // Check args are equal length arrays
+        with_attr error_message("ERC1155: accounts and ids length mismatch") {
+            assert ids_len = accounts_len;
+        }
+        // Allocate memory
+        let (local batch_balances: Uint256*) = alloc();
+        // Call iterator
+        _balance_of_batch_iter(accounts_len, accounts, ids, batch_balances);
+        return (accounts_len, batch_balances);
+    }
+
+    func is_approved_for_all{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        account: felt, operator: felt
+    ) -> (approved: felt) {
+        let (approved) = ERC1155_operator_approvals.read(account, operator);
+        return (approved,);
+    }
+
+    //
+    // Externals
+    //
+
+    func set_approval_for_all{syscall_ptr: felt*, pedersen_ptr: HashBuiltin*, range_check_ptr}(
+        operator: felt, approved: felt
+    ) {
+        let (caller) = get_caller_address();
+        with_attr error_message("ERC1155: cannot approve from the zero address") {
```

Comment:
accounts can be 1155

Reply:
i think contracts counterfactually deployed might also be susceptible to this if they use these methods inside `__validate_deploy__`

Reply:
If an Account is an 1155, calling itself will change the `caller_address` to itself (using `call_contract` syscall, I need to confirm this but this is the expected behavior IMO not being the call JUMP).

The only possibility I can think of for this to be possible (zero address caller) is if the Account being 1155 directly call (JUMP) these methods on the __execute__ entrypoint, and then only the Account deployer (who is the 1155 deployer) can "call" from zero address (bad or malicious design). I'm not sure how vital gas optimizations are here (compared to Ethereum for example), but I don't see a necessity to enforce this check when usually accounts won't be 1155 and it the case they are, this won't break anything because zero address can't receive funds anyway (`_safe_transfer_from` always revert on zero address target).

__validate_deploy__ jumping to a transfer method of an 1155 account is also a malicious or really poor design, and I don't think we need to enforce this check because of that (keeping also in mind that zero address balance will be 0)

Reply:
although it is true **our** account model forces all account function calls to be through `__execute__`, effectively setting a caller ≠ 0, it could be the case where someone develops a custom account without following our pattern, but still using our 1155 token. so yes, it could be the case that `__execute__` directly calls an 1155 function.

Reply:
although it is true **our** account model forces all account interactions through `__execute__`, custom account implementers may not follow our pattern but still use our 1155 module. so even if we agree it's bad design, it is possible for at least `__execute__` and `constructor` to be called from zero. i'm not that sure about `__validate__` and `__validate_deploy__` considering their limitations.

> I don't see a necessity to enforce this check when usually accounts won't be 1155 

as a security library we need to make our api as hard as misuse as possible, with special attention to edge cases.

> this won't break anything because zero address can't receive funds anyway (_safe_transfer_from always revert on zero address target)

this is true, but the set approval function would succeed and emit a falsy event.

Reply:
> although it is true **our** account model forces all account interactions through `__execute__`, custom account implementers may not follow our pattern but still use our 1155 module

Who is forcing the `__execute__` is the protocol itself, not our account right? Custom account implementors would still need to follow this.

> as a security library we need to make our api as hard as misuse as possible, with special attention to edge cases.

Agree, but we would be making transactions more expensive trying to avoid a case that would be a misuse of the protocol developer, not the user, nor us. `__execute__` has a well-defined purpose, and if a developer tries to redefine this purpose to JUMP to an ERC1155 method (as I said if Account is ER1155 but it calls, not JUMP, it won't be a problem), in that design he needs to manage edge cases, because he is redefining the purpose.

I feel optimizations are super important, but that's because my Solidity background, perhaps in this case is not that important.

Reply:
Hey, hope you don't mind me popping in to this discussion, but we had the same argument in a different project I'm part of. For what it's worth, having the `assert_not_zero` check adds only 4 steps to the TX cost (no builtins). That's it.

We opted into keeping it then.

Reply:
> Who is forcing the `__execute__` is the protocol itself, not our account right? Custom account implementors would still need to follow this.

 actually we enforced that first, then the protocol forced it. but even if the protocol forces you to go through `__execute__`, you can still add 1155 functions in its body, departing from our model that no function of the account will be called in there.

> Agree, but we would be making transactions more expensive by default

this is what we do everywhere, even in Solidity. we add checks to protect from misusage, which adds costs.

> trying to avoid a case that would be a misuse of the protocol developer, not the user, nor us.

our user is the smart contract developer, it's the misusage we care about.

> (and my feeling about how often this will happen is less than 0.1 percent, to be conservative, much less in my opinion).

you can't build a security lib by just focusing on the 99.9% of the happy paths. an attacker will analyze a system to find that 0.1%. this mindset is not negotiable.

Reply:
> having the `assert_not_zero` check adds only 4 steps to the TX cost (no builtins).

this is vey useful to know. thanks!

Reply:
> this is what we do everywhere, even in Solidity. we add checks to protect from misusage, which adds costs.

Yes, but there is a line. We don't check in Solidity if a Governor contract is also an ERC20Votes and uses himself to count votes, even when this is possible, because this is more in the scope of an audit than a library check, that's my feeling.

But following up on Milan comment, I agree that does no harm to keep the check.

Reply:
In OpenZeppelin Contracts for Solidity we would add the check to make sure the argument isn't 0. I see no reason to expect someone would call the function through `__execute__`.

Reply:
thanks for your input @frangio @milancermak.

i think this boils down to three different items/discussions:
e
1) the design principle we abide by ("make it very very difficult to do insecure stuff")
2) recognizing the fact that from-zero calls can happen in many context (thus apply point `1`)
3) whether we should add this check

i'm very much

Reply:
thanks for your input @frangio @milancermak. i think this is settled now, but i'd still like to boil this down to three different points for alignment:

1) agreeing on the design principle we abide by: to make it very very difficult to do insecure stuff with our tools
2) recognizing the fact that from-zero calls can happen in many context, thus apply point (1)
3) based on (2), we'll add this specific check not strictly for security but to avoid emitting a false event